File: examples/11-npm-local/tests/testthat/test-page.R
# Pins this example's distinctive `(test)` leaf for app.R: the page ships no
# shinyreact JS. The client runtime lives entirely in www/ui.js, bundled from
# `@posit-dev/shinyreact`, so the page must carry this app's dependency and
# only this app's — and no #shinyreact-config tag, which page_bare() is alone
# in never emitting.
#
# Mirrors Python's tests/test_page.py. It rebuilds app.R's two-line ui rather
# than sourcing the app, because sourcing it would start a Shiny app.
#
# Run it from the app directory:
#
#   Rscript -e 'shiny::runTests()'

library(shinyreact)

# The claim under test is how the page is composed, not what www/ holds, so the
# ui is built against an empty asset directory in tempdir(). That keeps the
# test working from the app directory AND from pkg-r's suite, which sources it
# with a different working directory. page_react_dep() requires the directory
# to exist (#290) but only warns about the missing ui.js inside it.
app_ui <- function() {
  dir <- file.path(tempdir(), "npm-local-page-test")
  dir.create(file.path(dir, "www"), recursive = TRUE, showWarnings = FALSE)
  old <- setwd(dir)
  on.exit(setwd(old), add = TRUE)

  suppressWarnings(page_bare(
    page_react_dep("www", name = "npm-local"),
    title = "Old Faithful"
  ))
}

test_that("the page carries only this app's dependency", {
  ui <- app_ui()
  names <- vapply(htmltools::findDependencies(ui), function(d) d$name, "")

  expect_true("npm-local" %in% names)
  expect_false("shinyreact" %in% names)
})

test_that("the page has no config tag", {
  ui <- app_ui()

  expect_false(grepl(
    "shinyreact-config",
    paste(as.character(htmltools::renderTags(ui)$html), collapse = "")
  ))
})

File: examples/11-npm-local/tests/test_page.py
"""Pins this example's distinctive `(test)` leaf: the page ships no shinyreact JS.

The client runtime lives entirely in `www/ui.js`, bundled from
`@posit-dev/shinyreact`. So the page must carry this app's dependency and
*only* this app's — no `shinyreact.js`, and no `#shinyreact-config` tag, which
`page_bare()` is alone in never emitting.

Mirrored in `tests/testthat/test-page.R` for `app.R`. Like that file, it
rebuilds `app.py`'s two-line ui against an empty asset directory rather than
importing the app: the claim is about how the page is composed, not about what
`www/` holds, and importing `app.py` would trigger its build-on-first-run.

Run it from the app directory, the way a user of the app would::

    pytest
"""

from __future__ import annotations

from pathlib import Path

import pytest
from shinyreact import page_bare, page_react_dep


@pytest.fixture
def ui(tmp_path: Path):
    # page_react_dep() requires the directory to exist, and only warns about
    # the missing ui.js inside it.
    (tmp_path / "www").mkdir()
    with pytest.warns(UserWarning, match="JS entry point not found"):
        return page_bare(
            page_react_dep(src_dir=tmp_path / "www", name="npm-local"),
            title="Old Faithful",
        )


def test_the_page_carries_only_this_apps_dependency(ui):
    names = {dep.name for dep in ui.get_dependencies()}
    assert "npm-local" in names
    assert "shinyreact" not in names


def test_the_page_has_no_config_tag(ui):
    assert "shinyreact-config" not in str(ui)
